WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldsListViewModel.swift

```diff
@@ -157,4 +208,50 @@ extension CustomFieldsListViewModel {
             self.fieldId = customField.id
         }
     }
+
+    struct PendingCustomFieldsChanges {
+        let editedFields: [CustomFieldUI]
+        let addedFields: [CustomFieldUI]
+        let deletedFieldIds: [Int64]
+
+        var hasChanges: Bool {
+            editedFields.isNotEmpty || addedFields.isNotEmpty || deletedFieldIds.isNotEmpty
+        }
+
+        init(editedFields: [CustomFieldUI] = [],
+             addedFields: [CustomFieldUI] = [],
+             deletedFieldIds: [Int64] = []) {
+            self.editedFields = editedFields
+            self.addedFields = addedFields
+            self.deletedFieldIds = deletedFieldIds
+        }
+
+        func copy(editedFields: [CustomFieldUI]? = nil,
+                  addedFields: [CustomFieldUI]? = nil,
+                  deletedFieldIds: [Int64]? = nil) -> PendingCustomFieldsChanges {
+            PendingCustomFieldsChanges(editedFields: editedFields ?? self.editedFields,
+                                       addedFields: addedFields ?? self.addedFields,
+                                       deletedFieldIds: deletedFieldIds ?? self.deletedFieldIds)
+        }
+
+        func asJson() -> [[String: Any?]] {
+            func metaDataAsJson(_ field: CustomFieldUI) -> [String: Any] {
+                var json: [String: Any] = [:]
+                if let fieldId = field.fieldId {
+                    json["id"] = fieldId
+                }
+                json["key"] = field.key
+                json["value"] = field.value
+                return json
+            }
```

Comment:
Personally I'm a fan of nested functions (local functions in Kotlin) when the logic is not to be used elsewhere, it keeps the logic in the same place, and it's easier to read, so I prefer to keep it like this, unless it's a blocking point.

Reply:
It's not quite common to use nested functions in iOS (at least from what I see). In iOS we tend to keep functions short with a single purpose. We try to avoid nested closures as well, as the code is harder to parse with those. It's one of the motivations for the async-await syntax as well - to avoid the nested hell of DispatchQueue. With helper methods that are not used elsewhere, I'd create a private extension for the related type to keep things separate.

I understand there can be different ideals between different languages and development teams. I'm good if you prefer to keep it your way - just wanna share the context of my suggestions. I hope it's helpful.

Reply:
I think it's a subjective topic, and it's not about the language itself, from my experience (in both Kotlin and C# before) and from what I already see shared ([for example](https://stackoverflow.com/a/41984023)), many people (including myself) like the readability advantages of nested functions 🙂.

Anyway, le'ts not drag this more, I updated the code to use an instance function in [9bdaa1a](https://github.com/woocommerce/woocommerce-ios/pull/14194/commits/9bdaa1a04bf2c981b1dd7d88c11500ba484807ec) 🙂